node/core/approval-voting/src/ops.rs

```diff
@@ -0,0 +1,386 @@
+					e.insert(match overlayed_db.load_candidate_entry(candidate_hash)? {
+						None => continue, // Should not happen except for corrupt DB
+						Some(c) => c,
+					})
+				}
+			};
+
+			candidate.block_assignments.remove(&block_hash);
+		}
+
+		Ok(block_entry.children)
+	};
+
+	// First visit everything before the height.
+	for i in range.0..canon_number {
+		let at_height = overlay_db.load_blocks_at_height(&i)?;
+		overlay_db.delete_blocks_at_height(i);
+
+		for b in at_height {
+			let _ = visit_and_remove_block_entry(
+				b,
+				overlay_db,
+				&mut visited_candidates,
+			)?;
+		}
+	}
+
+	// Then visit everything at the height.
+	let pruned_branches = {
+		let at_height = overlay_db.load_blocks_at_height(&canon_number)?;
+		overlay_db.delete_blocks_at_height(canon_number);
+
+		// Note that while there may be branches descending from blocks at earlier heights,
+		// we have already covered them by removing everything at earlier heights.
+		let mut pruned_branches = Vec::new();
+
+		for b in at_height {
+			let children = visit_and_remove_block_entry(
+				b,
+				overlay_db,
+				&mut visited_candidates,
+			)?;
+
+			if b != canon_hash {
+				pruned_branches.extend(children);
+			}
+		}
+
+		pruned_branches
+	};
+
+	// Follow all children of non-canonicalized blocks.
+	{
+		let mut frontier: Vec<(BlockNumber, Hash)> = pruned_branches.into_iter().map(|h| (canon_number + 1, h)).collect();
+		while let Some((height, next_child)) = frontier.pop() {
+			let children = visit_and_remove_block_entry(
+				next_child,
+				overlay_db,
+				&mut visited_candidates,
+			)?;
+
+			// extend the frontier of branches to include the given height.
+			frontier.extend(children.into_iter().map(|h| (height + 1, h)));
+
+			// visit the at-height key for this deleted block's height.
+			let at_height = match visited_heights.entry(height) {
+				Entry::Occupied(e) => e.into_mut(),
+				Entry::Vacant(e) => e.insert(overlay_db.load_blocks_at_height(&height)?),
+			};
+			if let Some(i) = at_height.iter().position(|x| x == &next_child) {
+				at_height.remove(i);
+			}
+		}
+	}
+
+	// Update all `CandidateEntry`s, deleting all those which now have empty `block_assignments`.
+	for (candidate_hash, candidate) in visited_candidates.into_iter() {
+		if candidate.block_assignments.is_empty() {
+			overlay_db.delete_candidate_entry(&candidate_hash);
+		} else {
+			overlay_db.write_candidate_entry(candidate);
+		}
+	}
+
+	// Update all blocks-at-height keys, deleting all those which now have empty `block_assignments`.
+	for (h, at) in visited_heights.into_iter() {
+		if at.is_empty() {
+			overlay_db.delete_blocks_at_height(h);
+		} else {
+			overlay_db.write_blocks_at_height(h, at);
+		}
+	}
+
+	// due to the fork pruning, this range actually might go too far above where our actual highest block is,
+	// if a relatively short fork is canonicalized.
+	let new_range = StoredBlockRange(
+		canon_number + 1,
+		std::cmp::max(range.1, canon_number + 2),
```

Comment:
I don't quite get this line? Should we not error out if the fork is of `0` length?

Reply:
Good question. This was previously in the approval_db/v1/mod.rs... I'm also not sure how to handle this properly. Currently, we never store None for the StoredBlockRange; it's only ever None on initialization. Will need to think about this more, marking as TBD.

Reply:
Adding a `TODO https://github.com/#issue` would be the very least. If it is important to know whether it was initialized, we should introduce a tristate enum - I think the above is just an unnecessary shortcut.

Reply:
That's a good point. Added. Will revisit with Rob to determine best path forward. Thanks for the encouraging commentary as well.

Reply:
If I recall correctly, the issue was that we might have two forks

A <- B1 <- C1 <- D1
A <- B2

and if B2 is finalized, the `StoredBlockRange` will still store the height of `D1 + 1` as the highest block even though we've pruned the entire B1 fork. It's difficult to improve this while still making a single atomic write and without adding another index.